### PR TITLE
fix: Component.propagateToChildren() order of traversal

### PR DESCRIPTION
**Pull request by st-pasha**
# Description

* `propagateToChildren` is supposed to visit components in the order opposite of how they were rendered. However, there was a bug that when called with `includeSelf=true`, the root of the tree will be visited first instead of last.
* Also, there was a bug where, in case of a game-in-game structure, the children of the nested game would be traversed twice.
* Added flag `bool reversed = false` to the `descendants()` method.

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [-] I have updated/added relevant examples in `examples`.

## Breaking Change

- [-] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

## Related Issues

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

**Review by spydon — COMMENTED** [review]
> However, there was a bug that when called with includeSelf=true, the root of the tree will be visited first instead of last.

Isn't it supposed to be visited first though? Since it is rendered and updated before its children?

File: packages/flame/lib/src/components/component.dart [hunk 1]
```diff
@@ -331,26 +331,35 @@ class Component {
...
-  /// words, it explores the first child completely before visiting the next
-  /// sibling.
-  ///
-  /// In order to filter descendants it is usually convenient to use [Iterable]
-  /// lazy methods, such as [Iterable.where].
-  Iterable<Component> descendants({bool includeSelf = false}) sync* {
-    if (includeSelf) {
+  /// The search is depth-first in preorder. In other words, it explores the
+  /// first child completely before visiting the next sibling, and the root
+  /// component is visited before its children.
```
Comment (spydon) [inline packages/flame/lib/src/components/component.dart:336]: This sounds like BFS to me, not DFS?

File: packages/flame/lib/src/components/component.dart [hunk 2]
```diff
@@ -331,26 +331,35 @@ class Component {
...
+  /// on every game cycle. The optional parameter [reversed] allows iterating
+  /// through the same set of descendants in reverse order.
+  ///
+  /// The [Iterable] produced by this method is "lazy", which means it will only
+  /// traverse the component tree when required. This allows efficient chaining
+  /// of various iterable methods, such as filtering, early stopping, folding,
+  /// and so on -- see the documentation of the [Iterable] class for details.
+  Iterable<Component> descendants({
+    bool includeSelf = false,
+    bool reversed = false,
```
Comment (spydon) [inline packages/flame/lib/src/components/component.dart:349]: Maybe we should have this flag on ancestors too?

**Review by spydon — APPROVED** [review]

**Review by erickzanardo — APPROVED** [review]

File: packages/flame/lib/src/components/component.dart (hunk 2 above)
Comment (st-pasha) [reply packages/flame/lib/src/components/component.dart:349]: I don't think there is a way to implement this for ancestors more efficiently than `ancestors().toList().reversed`.

Comment (spydon) [reply packages/flame/lib/src/components/component.dart:349]: True, since we don't have access to the root that it is attached to. Let's skip that then.

**Pull request merged**